File: AADS/self_test/quick_sort.py
# in place
# O(nlog(n))
import random
import math

def quick_sort(l, start=0, end=None):
  if end is None:
    end = len(l)

  if end <= start:
    return


  pivot_i = math.floor((start + end) / 2)
  pivot = l[pivot_i]
  switch(l, start, pivot_i)
  low_i = start + 1

  for i in range(start + 1, end):

    if l[i] < pivot:
      switch(l, i, low_i)
      low_i += 1

  new_pivot_i = low_i - 1
  switch(l, start, new_pivot_i)
  quick_sort(l, start, new_pivot_i)
  quick_sort(l, new_pivot_i + 1, end)


def switch(l, i, k):
  temp = l[i]
  l[i] = l[k]
  l[k] = temp


a = list(range(0, 300))
random.shuffle(a)
quick_sort(a)
print(a)

File: AADS/self_test/quickSort.js
const shuffle = (array) => {
  for (var i = array.length - 1; i > 0; i--) {
    var j = Math.floor(Math.random() * (i + 1));
    var temp = array[i];
    array[i] = array[j];
    array[j] = temp;
  }
}

const pickPivot = (start, end) => Math.floor((start + end) / 2)

const swap = (array, i, k) => {
  temp = array[i]
  array[i] = array[k]
  array[k] = temp
}

const partition = (array, start, end, pivotI) => {
  pivotValue = array[pivotI]
  swap(array, start, pivotI)
  lowI = start + 1

  for (i = start + 1; i <= end; i++) {
    if (array[i] < pivotValue) {
      swap(array, i, lowI)
      lowI++
    }
  }

  finalPivotI = lowI - 1
  swap(array, start, finalPivotI)

  return finalPivotI
}

const quickSort = (array, start=0, end) => {
  if (isNaN(end)) end = array.length - 1
  if (start >= end) return

  pivotI = pickPivot(start, end)
  finalPivotI = partition(array, start, end, pivotI)
  quickSort(array, start, finalPivotI)
  quickSort(array, finalPivotI + 1, end)
}

const a = [...Array(301).keys()]
shuffle(a)
console.log(a)
quickSort(a)
console.log(a)
